test/api/route/messages_test.dart

```diff
@@ -436,4 +453,195 @@ void main() {
+      required Anchor anchor,
+      required int numBefore,
+      required int numAfter,
+      required ApiNarrow narrow,
+      required UpdateMessageFlagsOp op,
+      required MessageFlag flag,
+      required Map<String, String> expected,
+    }) async {
+      final result = await updateMessageFlagsForNarrow(connection,
+        anchor: anchor, numBefore: numBefore, numAfter: numAfter,
+        narrow: narrow, op: op, flag: flag);
+      check(connection.lastRequest).isNotNull().isA<http.Request>()
+        ..method.equals('POST')
+        ..url.path.equals('/api/v1/messages/flags/narrow')
+        ..bodyFields.deepEquals(expected);
+      return result;
+    }
+    final fakeResult = UpdateMessageFlagsForNarrowResult(
+      processedCount: 11, updatedCount: 11, foundOldest: true,
+      foundNewest: true);
+
+    test('smoke', () {
+      return FakeApiConnection.with_((connection) async {
+        connection.prepare(json: fakeResult.toJson());
+        await checkUpdateMessageFlagsForNarrow(connection,
+          anchor: AnchorCode.oldest,
+          numBefore: 0, numAfter: 20,
+          narrow: const AllMessagesNarrow().apiEncode(),
+          op: UpdateMessageFlagsOp.add, flag: MessageFlag.read,
+          expected: {
+            'anchor': 'oldest',
+            'num_before': '0',
+            'num_after': '20',
+            'narrow': jsonEncode([]),
+            'op': 'add',
+            'flag': 'read',
+          });
+      });
+    });
+
+    test('numeric anchor', () {
+      return FakeApiConnection.with_((connection) async {
+        connection.prepare(json: fakeResult.toJson());
+        await checkUpdateMessageFlagsForNarrow(connection,
+          anchor: const NumericAnchor(42),
+          numBefore: 0, numAfter: 20,
+          narrow: const AllMessagesNarrow().apiEncode(),
+          op: UpdateMessageFlagsOp.add, flag: MessageFlag.read,
+          expected: {
+            'anchor': '42',
+            'num_before': '0',
+            'num_after': '20',
+            'narrow': jsonEncode([]),
+            'op': 'add',
+            'flag': 'read',
+          });
+      });
+    });
+
+    test('narrow uses resolveLegacyElements to encode', () {
+      return FakeApiConnection.with_((connection) async {
+        connection.zulipFeatureLevel = 176;
+        connection.prepare(json: fakeResult.toJson());
+        await checkUpdateMessageFlagsForNarrow(connection,
+          anchor: AnchorCode.oldest,
+          numBefore: 0, numAfter: 20,
+          narrow: [ApiNarrowDm([123, 234])],
+          op: UpdateMessageFlagsOp.add, flag: MessageFlag.read,
+          expected: {
+            'anchor': 'oldest',
+            'num_before': '0',
+            'num_after': '20',
+            'narrow': jsonEncode([
+              {'operator': 'pm-with', 'operand': [123, 234]},
+            ]),
+            'op': 'add',
+            'flag': 'read',
+          });
+        connection.zulipFeatureLevel = eg.futureZulipFeatureLevel;
```

Comment:
Let's take this line out; I feel like it's likely to just confuse someone about the semantics of `FakeApiConnection.with_`, which are that it's a fresh connection that's thrown away.